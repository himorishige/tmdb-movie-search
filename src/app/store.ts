import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import userReducer from 'src/features/user/userSlice';
import { setupListeners } from '@reduxjs/toolkit/query';
import { movieApi } from 'src/features/movie/movieQuery';

export const store = configureStore({
  reducer: {
    user: userReducer,
    [movieApi.reducerPath]: movieApi.reducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(movieApi.middleware),
});

setupListeners(store.dispatch);

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
