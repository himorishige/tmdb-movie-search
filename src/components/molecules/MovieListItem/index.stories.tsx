import { Story, Meta } from '@storybook/react';
import MovieListItem, { Props } from '.';

export default {
  title: 'TMDB Search/Molecules/MovieListItem',
  component: MovieListItem,
} as Meta;

const Template: Story<Props> = (args) => <MovieListItem {...args} />;

export const Default = Template.bind({});
Default.args = {
  data: {
    adult: false,
    backdrop_path: '/nVKRspU9SQEs2gNrms8cDKsI4vx.jpg',
    genre_ids: [28, 878, 12, 10752],
    id: 588228,
    original_language: 'en',
    original_title: 'The Tomorrow War',
    overview: '',
    poster_path: '/xipF6XqfSYV8DxLqfLN6aWlwuRp.jpg',
    release_date: '2021-06-30',
    title: 'トゥモロー･ウォー',
    video: false,
    vote_average: 8.7,
    vote_count: 695,
    popularity: 771.266,
  },
};
