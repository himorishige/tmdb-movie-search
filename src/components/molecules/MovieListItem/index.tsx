import { useState, useRef, memo, useCallback } from 'react';
import { Link as RouterLink } from 'react-router-dom';
import { Box, Flex, GridItem, Heading, Image, Text, Link } from '@chakra-ui/react';
import { DetailResponse, MovieData } from 'src/types';
import { IMAGE_BASE_URL } from 'src/constants';
import { useHelper } from 'src/hooks/useHelper';
import { useAppDispatch, useAppSelector } from 'src/app/hooks';
import { toggleFavoritesList, selectFavoritesIds } from 'src/features/user/userSlice';
import { useToastMessage } from 'src/hooks/useToastMessage';
import { FavoriteButton } from 'src/components/atoms';
import { AlertDialogBox } from 'src/components/organisms';

export type Props = {
  data: MovieData | DetailResponse;
};

const MovieListItem: React.VFC<Props> = memo(({ data }) => {
  const dispatch = useAppDispatch();
  const ids = useAppSelector(selectFavoritesIds);
  const favoriteFlag = ids.includes(data.id);
  const { formatLocalDate } = useHelper();
  const { showMessage } = useToastMessage();

  // お気に入りの削除アラート制御
  const [isOpen, setIsOpen] = useState(false);
  const onClose = useCallback(() => setIsOpen(false), []);
  const cancelRef = useRef().current;

  // お気に入りボタンのトグル制御
  const toggleHandler = useCallback(() => {
    if (favoriteFlag) {
      setIsOpen(true);
    } else {
      dispatch(toggleFavoritesList(data));
      showMessage({ title: 'お気に入りリストに登録しました。', status: 'success' });
    }
  }, [data, dispatch, favoriteFlag, showMessage]);

  // お気に入りリストから削除する制御
  const removeHandler = useCallback(() => {
    dispatch(toggleFavoritesList(data));
    setIsOpen(false);
    showMessage({ title: '削除に成功しました。', status: 'success' });
  }, [data, dispatch, showMessage]);

  return (
    <GridItem>
      <Flex direction="column" alignItems="center">
        <Box>
          <Box position="relative">
            <Flex
              position="absolute"
              bottom="0"
              right="0"
              bgColor="whiteAlpha.500"
              w={16}
              h={16}
              clipPath="polygon(0 100%, 100% 0, 100% 100%)"
              alignItems="flex-end"
              justifyContent="flex-end"
              p={0}
            >
              <FavoriteButton favoriteFlag={favoriteFlag} toggleHandler={toggleHandler} />
            </Flex>
            <Link as={RouterLink} to={`/movie/${data.id}`}>
              <Image
                w="auto"
                h="auto"
                maxH="270px"
                objectFit="contain"
                src={`${IMAGE_BASE_URL}${data.poster_path}`}
                fallbackSrc="/assets/no-image.jpg"
                alt={data.title}
                borderRadius={8}
                bgColor="gray.100"
              />
            </Link>
          </Box>
        </Box>
        <Box p={2} w="100%">
          <Heading size="s">
            <Link as={RouterLink} to={`/movie/${data.id}`}>
              {data.title}
            </Link>
          </Heading>
          <Text>{formatLocalDate(data.release_date)}</Text>
        </Box>
      </Flex>

      {/* 削除確認用のアラート表示 */}
      <AlertDialogBox
        title={data.title}
        isOpen={isOpen}
        cancelRef={cancelRef}
        onClose={onClose}
        removeHandler={removeHandler}
      />
    </GridItem>
  );
});

export default MovieListItem;
