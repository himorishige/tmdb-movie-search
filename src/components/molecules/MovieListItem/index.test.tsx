import renderer from 'react-test-renderer';
import MovieListItem, { Props } from '.';
import { store } from 'src/app/store';
import { Provider } from 'react-redux';
import { HelmetProvider } from 'react-helmet-async';
import { ChakraProvider } from '@chakra-ui/react';
import { BrowserRouter } from 'react-router-dom';

describe('Molecules/MovieListItem', () => {
  test('正しくレンダリングされている', () => {
    const props: Props = {
      data: {
        adult: false,
        backdrop_path: '/nVKRspU9SQEs2gNrms8cDKsI4vx.jpg',
        genre_ids: [28, 878, 12, 10752],
        id: 588228,
        original_language: 'en',
        original_title: 'The Tomorrow War',
        overview: '',
        poster_path: '/xipF6XqfSYV8DxLqfLN6aWlwuRp.jpg',
        release_date: '2021-06-30',
        title: 'トゥモロー･ウォー',
        video: false,
        vote_average: 8.7,
        vote_count: 695,
        popularity: 771.266,
      },
    };
    const element = renderer
      .create(
        <HelmetProvider>
          <Provider store={store}>
            <ChakraProvider>
              <BrowserRouter>
                <MovieListItem {...props} />
              </BrowserRouter>
            </ChakraProvider>
          </Provider>
        </HelmetProvider>,
      )
      .toJSON();
    expect(element).toMatchSnapshot();
  });
});
