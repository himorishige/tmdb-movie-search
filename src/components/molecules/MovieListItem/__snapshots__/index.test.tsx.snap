// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecules/MovieListItem 正しくレンダリングされている 1`] = `
Array [
  <div
    className="css-0"
  >
    <div
      className="css-dvxtzn"
    >
      <div
        className="css-0"
      >
        <div
          className="css-79elbk"
        >
          <div
            className="css-n0d1fk"
          >
            <button
              aria-label="remove FavoriteList"
              className="chakra-button css-1lppsyo"
              data-testid="add-favorites-button"
              onClick={[Function]}
              type="button"
            >
              <svg
                aria-hidden={true}
                fill="currentColor"
                focusable={false}
                fontSize="1rem"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
                />
              </svg>
            </button>
          </div>
          <a
            className="chakra-link css-f4h6uy"
            href="/movie/588228"
            onClick={[Function]}
          >
            <img
              alt="トゥモロー･ウォー"
              className="chakra-image__placeholder css-1435m0h"
              src="/assets/no-image.jpg"
            />
          </a>
        </div>
      </div>
      <div
        className="css-9elzxb"
      >
        <h2
          className="chakra-heading css-1eam21a"
        >
          <a
            className="chakra-link css-f4h6uy"
            href="/movie/588228"
            onClick={[Function]}
          >
            トゥモロー･ウォー
          </a>
        </h2>
        <p
          className="chakra-text css-0"
        >
          2021年06月30日
        </p>
      </div>
    </div>
  </div>,
  <span />,
]
`;
